### DISCUSSION TITLE
Proxy Server Support?

### DISCUSSION THREAD

**Discussion by Frogley** [betalgo/openai#148] [Q&A] [answered]
In my region(mainland China), I cannot access the OPENAI API directly and must use a proxy server. Could we add support for proxy servers? Here's what the code looks like.

> OpenAI.GPT3/Managers/OpenAIService.cs

```
  if (httpClient != null)
  {
      _httpClient = httpClient;
  }
  else
  {
      if (!string.IsNullOrEmpty(settings.Proxy))
      {
          var proxy = new WebProxy
          {
              Address = new Uri(settings.Proxy),
              BypassProxyOnLocal = false,
              UseDefaultCredentials = false,
          };
          // Create a client handler that uses the proxy
          var httpClientHandler = new HttpClientHandler
          {
              Proxy = proxy,
          };
          // Disable SSL verification
          httpClientHandler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
          // Finally, create the HTTP client object
          _httpClient = new HttpClient(handler: httpClientHandler, disposeHandler: true);
      }
      else
      {
          _httpClient = new HttpClient();
      }
  }

```
> Example to user proxy Server
```
  var openAiService = new OpenAIService(new OpenAiOptions()
  {
      ApiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY") ?? "",
      Proxy = "Proxy Server Address",
  });
```
Of course, this is just the simplest scenario. 

What do you think about this?

**Comment by kayhantolga** [comment]
Additional support may be added in the future, but for now, I think you can achieve your goal with the current code. Check the constructor method where you can pass your HTTP client. Although it's a bit limited at the moment, it may solve your issue. Also, you can set the BaseDomain to use a custom domain.
Please let me know if this solution would be helpful for you.

https://github.com/betalgo/openai/blob/master/OpenAI.SDK/Managers/OpenAIService.cs
Line 20
`public OpenAIService(OpenAiOptions settings, HttpClient? httpClient = null)`

**Reply by Frogley** [reply]
Actually I'm using the code above right now and it's working well. Thanks for your help.

**Reply by GrowingHacker** [reply]
你不用数据加密吗？

**Comment by tianqingshuilan** [comment]
> In my region(mainland China), I cannot access the OPENAI API directly and must use a proxy server. Could we add support for proxy servers? Here's what the code looks like.
> 
> > OpenAI.GPT3/Managers/OpenAIService.cs
> 
> ```
>   if (httpClient != null)
>   {
>       _httpClient = httpClient;
>   }
>   else
>   {
>       if (!string.IsNullOrEmpty(settings.Proxy))
>       {
>           var proxy = new WebProxy
>           {
>               Address = new Uri(settings.Proxy),
>               BypassProxyOnLocal = false,
>               UseDefaultCredentials = false,
>           };
>           // Create a client handler that uses the proxy
>           var httpClientHandler = new HttpClientHandler
>           {
>               Proxy = proxy,
>           };
>           // Disable SSL verification
>           httpClientHandler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
>           // Finally, create the HTTP client object
>           _httpClient = new HttpClient(handler: httpClientHandler, disposeHandler: true);
>       }
>       else
>       {
>           _httpClient = new HttpClient();
>       }
>   }
> ```
> 
> > Example to user proxy Server
> 
> ```
>   var openAiService = new OpenAIService(new OpenAiOptions()
>   {
>       ApiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY") ?? "",
>       Proxy = "Proxy Server Address",
>   });
> ```
> 
> Of course, this is just the simplest scenario.
> 
> What do you think about this?

Hello fellow, if you need to use proxy in dependency injection, you can refer to my code.  
Of course, this is a webapi project, so there will be a WebApplicationBuilder link.

```CSharp
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddOptions<OpenAiOptions>().Configure(settings => { settings.ApiKey = "YOUR_KEY"; });

var webProxy = new WebProxy(AppSettingsConstVars.ProxyUrl, false);
var proxyHttpClientHandler = new HttpClientHandler
{
    Proxy = webProxy,
    UseProxy = AppSettingsConstVars.ProxyUseProxy
};
builder.Services.AddHttpClient<IOpenAIService, OpenAIService>(option =>
{
    option.Timeout = TimeSpan.FromSeconds(10);
}).ConfigurePrimaryHttpMessageHandler(() =>
    {
        return new HttpClientHandler { Proxy = "YOUR_PROXY_URL", UseProxy = true };
    });
```

**Comment by kayhantolga** [comment, marked as answer]
with the new version, you can use it like this
```csharp
serviceCollection.AddOpenAIService().ConfigurePrimaryHttpMessageHandler(() =>
    {
        return new HttpClientHandler { Proxy = "YOUR_PROXY_URL", UseProxy = true };
    });
```

**Reply by Frogley** [reply]
Got it, thanks.

**Reply by git102347501** [reply]
> 使用新版本，您可以像这样使用它
> 
> ```cs
> serviceCollection.AddOpenAIService().ConfigurePrimaryHttpMessageHandler(() =>
>     {
>         return new HttpClientHandler { Proxy = "YOUR_PROXY_URL", UseProxy = true };
>     });
> ```
I think this should be reflected in the md document of the demo to avoid more confusion

**Reply by phoben** [reply]
I don't have a proxy server, and I would like to use a mirror API provided by a third party as a substitute for OPENAI's address. How should I set this up?
